// organism-react-popup/build/cjs/src/ui/molecules/BasePopup.js
"use strict";

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("reshow-runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("reshow-runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("reshow-runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("reshow-runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("reshow-runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("reshow-runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _needCss = _interopRequireDefault(require("need-css"));

var _popupStore = require("../../stores/popupStore.js");

var BasePopup = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(BasePopup, _PureComponent);

  var _super = (0, _createSuper2["default"])(BasePopup);

  function BasePopup(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, BasePopup);
    _this = _super.call(this, props);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      hasError: false
    });
    (0, _needCss["default"])(["popup", "modal"], "semantic");
    return _this;
  }

  (0, _createClass2["default"])(BasePopup, [{
    key: "componentDidCatch",
    value: function componentDidCatch(error, info) {
      var onError = this.props.onError;

      if (onError) {
        callfunc(onError, [error, info]);
      } else {
        console.error([error, info]);
      }

      this.setState({
        hasError: true
      });
    }
  }, {
    key: "close",
    value: function close() {
      (0, _popupStore.popupDispatch)({
        type: "dom/closeOne",
        params: {
          popup: this
        }
      });
      return true;
    }
  }], [{
    key: "getDerivedStateFromError",
    value: function getDerivedStateFromError(error) {
      return {
        hasError: true
      };
    }
  }]);
  return BasePopup;
}(_react.PureComponent);

(0, _defineProperty2["default"])(BasePopup, "defaultProps", {
  name: "default"
});
var _default = BasePopup;
exports["default"] = _default;
module.exports = exports.default;
