// organism-react-popup/build/cjs/src/ui/molecules/PopupOverlay.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _classCallCheck2 = _interopRequireDefault(require("reshow-runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("reshow-runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("reshow-runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("reshow-runtime/helpers/createSuper"));

var _reshowReturn = _interopRequireDefault(require("reshow-return"));

var _reactAtomicMolecule = require("react-atomic-molecule");

var _getObjectValue = _interopRequireDefault(require("get-object-value"));

var _getStyle = _interopRequireDefault(require("get-style"));

var _BasePopup2 = _interopRequireDefault(require("../molecules/BasePopup.js"));

var _popupStore = _interopRequireWildcard(require("../../stores/popupStore.js"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["className"],
    _excluded2 = ["targetEl", "toPool", "alignParams", "retryAt", "isFollowTransform", "className", "style", "group"];

var PopupOverlay = /*#__PURE__*/function (_BasePopup) {
  (0, _inherits2["default"])(PopupOverlay, _BasePopup);

  var _super = (0, _createSuper2["default"])(PopupOverlay);

  function PopupOverlay() {
    (0, _classCallCheck2["default"])(this, PopupOverlay);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(PopupOverlay, [{
    key: "resetStyle",
    value: function resetStyle(key, thisStyle) {
      var _this = this;

      var value = (0, _getObjectValue["default"])(this.state, [key], function () {
        return (0, _getObjectValue["default"])(_this.props, [key]);
      });

      if ("undefined" !== typeof value) {
        thisStyle[key] = value + "px";
      }
    }
  }, {
    key: "renderOverlay",
    value: function renderOverlay(props) {
      var className = props.className,
          others = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);
      var classes = (0, _reactAtomicMolecule.mixClass)("popup", className);
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactAtomicMolecule.SemanticUI, (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, others), {}, {
        className: classes
      }));
    }
  }, {
    key: "shouldShow",
    value: function shouldShow(show) {
      var _this2 = this;

      if (!show) {
        return null;
      }

      var _this$props = this.props,
          targetEl = _this$props.targetEl,
          toPool = _this$props.toPool,
          alignParams = _this$props.alignParams,
          retryAt = _this$props.retryAt,
          isFollowTransform = _this$props.isFollowTransform,
          className = _this$props.className,
          style = _this$props.style,
          group = _this$props.group,
          others = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, _excluded2);
      /* <!-- Handle Style */

      var thisStyle = (0, _objectSpread2["default"])({}, style);
      this.resetStyle("top", thisStyle);
      this.resetStyle("left", thisStyle);
      this.resetStyle("width", thisStyle);
      this.resetStyle("height", thisStyle);

      if (targetEl && isFollowTransform) {
        thisStyle.transform = (0, _getStyle["default"])(targetEl, "transform");
      }

      others.style = thisStyle;
      /*  Handle Style --> */

      var refCb = (0, _getObjectValue["default"])(this.state, ["refCb"], function () {
        return (0, _getObjectValue["default"])(_this2.props, ["refCb"]);
      });

      if (refCb) {
        others.refCb = refCb;
      }

      others.className = (0, _reactAtomicMolecule.mixClass)(className, (0, _getObjectValue["default"])(this, ["state", "className"]), "visible");
      return this.renderOverlay(others);
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      /**
       * Use build + anonymous function to suppress render.
       */
      return this.state.hasError ? null : (0, _reactAtomicMolecule.build)(_reshowReturn["default"])({
        store: _popupStore["default"],
        initStates: [_popupStore.SHOW_KEY]
      }, function (props) {
        var name = _this3.props.name;
        var show = (0, _getObjectValue["default"])(props[_popupStore.SHOW_KEY], [name]);
        return _this3.shouldShow(show);
      });
    }
  }]);
  return PopupOverlay;
}(_BasePopup2["default"]);

var _default = PopupOverlay;
exports["default"] = _default;
module.exports = exports.default;
