// organism-react-popup/build/cjs/src/ui/organisms/FullScreen.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _react = _interopRequireWildcard(require("react"));

var _X = _interopRequireDefault(require("ricon/X"));

var _getObjectValue = _interopRequireDefault(require("get-object-value"));

var _callFunc = _interopRequireDefault(require("call-func"));

var _reactAtomicMolecule = require("react-atomic-molecule");

var _reshowHooks = require("reshow-hooks");

var _PopupModal = _interopRequireDefault(require("../molecules/PopupModal.js"));

var _DisplayPopupEl = _interopRequireDefault(require("../organisms/DisplayPopupEl.js"));

var _popupStore = require("../../stores/popupStore.js");

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["appear", "enter", "name", "xico", "scrolling", "page", "children", "className", "style"];

var DefaultXIcon = function DefaultXIcon(props) {
  var _useState = (0, _react.useState)(),
      hoverStyle = _useState[0],
      setHoverStyle = _useState[1];

  var _mounted = (0, _reshowHooks.useMounted)();

  var xIcoEnter = function xIcoEnter() {
    if (_mounted()) {
      setHoverStyle(Styles.xIcoHover);
    }
  };

  var xIcoLeave = function xIcoLeave() {
    if (_mounted()) {
      setHoverStyle(null);
    }
  };

  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_X["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, props), {}, {
    onMouseEnter: xIcoEnter,
    onMouseLeave: xIcoLeave,
    style: (0, _objectSpread2["default"])((0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, Styles.x), props.style), hoverStyle),
    size: "75px",
    weight: ".1rem"
  }));
};

var FullScreen = function FullScreen(_ref) {
  var _ref$appear = _ref.appear,
      appear = _ref$appear === void 0 ? "fadeIn-500" : _ref$appear,
      _ref$enter = _ref.enter,
      enter = _ref$enter === void 0 ? "fadeIn-500" : _ref$enter,
      _ref$name = _ref.name,
      name = _ref$name === void 0 ? "fullscreen" : _ref$name,
      _ref$xico = _ref.xico,
      xico = _ref$xico === void 0 ? DefaultXIcon : _ref$xico,
      _ref$scrolling = _ref.scrolling,
      scrolling = _ref$scrolling === void 0 ? false : _ref$scrolling,
      _ref$page = _ref.page,
      page = _ref$page === void 0 ? true : _ref$page,
      children = _ref.children,
      className = _ref.className,
      style = _ref.style,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(_ref, _excluded);
  var closeEl;
  var thisStyle = style;

  if (page) {
    closeEl = xico;
    thisStyle = (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, Styles.container), style);

    if (scrolling) {
      thisStyle.display = "block";
    }
  }

  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_DisplayPopupEl["default"], {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_PopupModal["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, restProps), {}, {
      name: name + " (modal)",
      className: (0, _reactAtomicMolecule.mixClass)("full-screen", className),
      style: thisStyle,
      modalStyle: Styles.modal,
      modal: children,
      closeEl: closeEl
    }))
  });
};

var _default = FullScreen;
exports["default"] = _default;
var Styles = {
  container: {
    background: "#fff",
    textAlign: "left",
    padding: 0
  },
  x: {
    width: "70px",
    height: "75px",
    borderRadius: "8px",
    backgroundColor: "rgba(190,190,190,.39)",
    top: "20px",
    right: "20px",
    opacity: ".3"
  },
  xIcoHover: {
    opacity: ".9"
  }
};
module.exports = exports.default;
