// organism-react-popup/build/cjs/src/ui/organisms/PopupClick.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _classCallCheck2 = _interopRequireDefault(require("reshow-runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("reshow-runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("reshow-runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("reshow-runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("reshow-runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("reshow-runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactAtomicMolecule = require("react-atomic-molecule");

var _getRandomId = require("get-random-id");

var _callFunc = _interopRequireDefault(require("call-func"));

var _getObjectValue = _interopRequireDefault(require("get-object-value"));

var _PopupOverlay = _interopRequireDefault(require("../molecules/PopupOverlay.js"));

var _DisplayPopupEl = _interopRequireDefault(require("../organisms/DisplayPopupEl.js"));

var _popupStore = require("../../stores/popupStore.js");

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["children", "style", "className", "component", "container", "popup", "callback", "onClose", "once"];

var PopupClick = /*#__PURE__*/function (_Component) {
  (0, _inherits2["default"])(PopupClick, _Component);

  var _super = (0, _createSuper2["default"])(PopupClick);

  function PopupClick() {
    var _this;

    (0, _classCallCheck2["default"])(this, PopupClick);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleClick", function () {
      return _this.open();
    });
    return _this;
  }

  (0, _createClass2["default"])(PopupClick, [{
    key: "open",
    value: function open() {
      var _this$props = this.props,
          callback = _this$props.callback,
          once = _this$props.once,
          popup = _this$props.popup;

      if (once) {
        (0, _popupStore.popupDispatch)({
          type: "dom/update",
          params: {
            popup: (0, _reactAtomicMolecule.build)(popup, {
              wrap: _PopupOverlay["default"],
              doCallFunction: true
            })()
          }
        });
        (0, _callFunc["default"])(callback);
      } else {
        this.setState({
          show: true,
          bust: (0, _getRandomId.getTimestamp)()
        }, function () {
          (0, _callFunc["default"])(callback);
        });
      }
    }
  }, {
    key: "close",
    value: function close() {
      var onClose = this.props.onClose;
      this.setState({
        show: false
      }, function () {
        (0, _callFunc["default"])(onClose);
      });
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var container = this.props.container;
      container && console.warn("Container will retire soon, change to use component");
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          children = _this$props2.children,
          style = _this$props2.style,
          className = _this$props2.className,
          component = _this$props2.component,
          container = _this$props2.container,
          popup = _this$props2.popup,
          callback = _this$props2.callback,
          onClose = _this$props2.onClose,
          once = _this$props2.once,
          reset = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, _excluded);

      var _ref = this.state || {},
          show = _ref.show,
          bust = _ref.bust;

      var thisStyle = (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, style), Styles.container);
      var popupEl = null;

      if (show && !once) {
        popupEl = /*#__PURE__*/(0, _jsxRuntime.jsx)(_DisplayPopupEl["default"], {
          bust: bust,
          children: (0, _reactAtomicMolecule.build)(popup, {
            wrap: _PopupOverlay["default"],
            doCallFunction: true
          })()
        }, "popup-el");
      }

      var thisComponent = component || container || _reactAtomicMolecule.SemanticUI;
      var thisChildren = [children || (0, _getObjectValue["default"])(thisComponent, ["props", "children"]), popupEl];
      var props = (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, reset), {}, {
        onClick: this.handleClick,
        className: (0, _reactAtomicMolecule.mixClass)(className, "popup-click"),
        style: thisStyle
      });
      return (0, _reactAtomicMolecule.build)(thisComponent)(props, thisChildren);
    }
  }]);
  return PopupClick;
}(_react.Component);

(0, _defineProperty2["default"])(PopupClick, "defaultProps", {
  once: true
});
var _default = PopupClick;
exports["default"] = _default;
var Styles = {
  container: {
    cursor: "pointer"
  }
};
module.exports = exports.default;
