// organism-react-popup/build/cjs/src/ui/organisms/PopupHover.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _classCallCheck2 = _interopRequireDefault(require("reshow-runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("reshow-runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("reshow-runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("reshow-runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("reshow-runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("reshow-runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _reactAtomicMolecule = require("react-atomic-molecule");

var _getRandomId = require("get-random-id");

var _callFunc = _interopRequireDefault(require("call-func"));

var _PopupFloatEl = _interopRequireDefault(require("../molecules/PopupFloatEl.js"));

var _DisplayPopupEl = _interopRequireDefault(require("../organisms/DisplayPopupEl.js"));

var _popupStore = require("../../stores/popupStore.js");

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["children", "popup", "isKeep", "toPool", "alignParams", "component", "name"];
var closeTimer = {};

var PopupHover = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(PopupHover, _PureComponent);

  var _super = (0, _createSuper2["default"])(PopupHover);

  function PopupHover() {
    var _this;

    (0, _classCallCheck2["default"])(this, PopupHover);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "floatMouseOver", function () {
      _this.isKeep = true;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "floatMouseOut", function () {
      _this.isKeep = false;

      _this.close();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "mouseOver", function () {
      var name = _this.props.name;
      clearTimeout(closeTimer[name]);

      _this.open();
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "mouseOut", function () {
      var name = _this.props.name;
      clearTimeout(closeTimer[name]);
      closeTimer[name] = setTimeout(function () {
        _this.close();
      }, 100);
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleDom", function (dom) {
      return _this.dom = dom;
    });
    return _this;
  }

  (0, _createClass2["default"])(PopupHover, [{
    key: "open",
    value: function open() {
      var callback = this.props.callback;
      this.setState({
        show: true,
        bust: (0, _getRandomId.getTimestamp)()
      }, function () {
        (0, _callFunc["default"])(callback);
      });
    }
  }, {
    key: "close",
    value: function close() {
      var _this$props = this.props,
          isKeep = _this$props.isKeep,
          onClose = _this$props.onClose;

      if (this.isKeep || isKeep) {
        return;
      }

      this.setState({
        show: false
      }, function () {
        (0, _callFunc["default"])(onClose);
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          children = _this$props2.children,
          popup = _this$props2.popup,
          isKeep = _this$props2.isKeep,
          toPool = _this$props2.toPool,
          alignParams = _this$props2.alignParams,
          component = _this$props2.component,
          name = _this$props2.name,
          others = (0, _objectWithoutPropertiesLoose2["default"])(_this$props2, _excluded);

      var _ref = this.state || {},
          show = _ref.show,
          bust = _ref.bust;

      var popupEl = null;

      if (show) {
        popupEl = /*#__PURE__*/(0, _jsxRuntime.jsx)(_DisplayPopupEl["default"], {
          bust: bust,
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_PopupFloatEl["default"], {
            targetEl: this.dom,
            toPool: toPool,
            name: name,
            alignParams: alignParams,
            onMouseEnter: this.floatMouseOver,
            onMouseLeave: this.floatMouseOut,
            children: popup
          })
        }, "popup-el");
      }

      var thisChildren = [children, popupEl];
      return (0, _reactAtomicMolecule.build)(component)((0, _objectSpread2["default"])({
        refCb: this.handleDom,
        onMouseEnter: this.mouseOver,
        onMouseLeave: this.mouseOut,
        name: name
      }, others), thisChildren);
    }
  }]);
  return PopupHover;
}(_react.PureComponent);

(0, _defineProperty2["default"])(PopupHover, "defaultProps", {
  name: "popup-hover",
  component: _reactAtomicMolecule.SemanticUI
});
var _default = PopupHover;
exports["default"] = _default;
module.exports = exports.default;
