// reshow/build/cjs/src/stores/clientStorageStore.js
"use strict";

exports.__esModule = true;
exports.sessionValueStore =
  exports.sessionValueDispatch =
  exports.sessionStorageStore =
  exports.sessionStorageDispatch =
  exports.localValueStore =
  exports.localValueDispatch =
  exports.localStorageStore =
  exports.localStorageDispatch =
    void 0;

var _reshowFluxBase = require("reshow-flux-base");

var _getStorage = require("get-storage");

var _winDoc = require("win-doc");

var isUpdating;
var lastDispatch = {
  current: {},
};

var getLocalReducer = function getLocalReducer(name, storage, disableEncode) {
  return (0, _reshowFluxBase.createReducer)(function (state, action) {
    isUpdating = true;
    lastDispatch.current = {
      name: name,
      action: action,
    };
    var result = state.merge(action);
    isUpdating = false;
    return result;
  }, new _getStorage.Storage(storage, disableEncode));
};

var LOCAL_STORAGE = "localStorage";
var LOCAL_VALUE = "localValue";
var SESSION_STORAGE = "sessionStorage";
var SESSION_VALUE = "sessionValue";

var _getLocalReducer = getLocalReducer(LOCAL_STORAGE, _getStorage.localStorage),
  localStorageStore = _getLocalReducer[0],
  localStorageDispatch = _getLocalReducer[1];

exports.localStorageDispatch = localStorageDispatch;
exports.localStorageStore = localStorageStore;

var _getLocalReducer2 = getLocalReducer(
    LOCAL_VALUE,
    _getStorage.localStorage,
    true
  ),
  localValueStore = _getLocalReducer2[0],
  localValueDispatch = _getLocalReducer2[1];

exports.localValueDispatch = localValueDispatch;
exports.localValueStore = localValueStore;

var _getLocalReducer3 = getLocalReducer(
    SESSION_STORAGE,
    _getStorage.sessionStorage
  ),
  sessionStorageStore = _getLocalReducer3[0],
  sessionStorageDispatch = _getLocalReducer3[1];

exports.sessionStorageDispatch = sessionStorageDispatch;
exports.sessionStorageStore = sessionStorageStore;

var _getLocalReducer4 = getLocalReducer(
    SESSION_VALUE,
    _getStorage.sessionStorage,
    true
  ),
  sessionValueStore = _getLocalReducer4[0],
  sessionValueDispatch = _getLocalReducer4[1];

exports.sessionValueDispatch = sessionValueDispatch;
exports.sessionValueStore = sessionValueStore;
