// reshow/build/es/src/ui/molecules/ReshowComponent.mjs
import _objectSpread from "reshow-runtime/es/helpers/objectSpread2";
import build from "reshow-build";
import { connectOptions, getReturn } from "reshow-return";
import { useConnect } from "reshow-flux";
import pageStore from "../../stores/pageStore.mjs";
import { globalStore } from "../../stores/globalStore.mjs";

var storeLocator = function storeLocator(props) {
  return (props === null || props === void 0 ? void 0 : props.store) || pageStore;
};

var initStates = ["data", "I18N"];
var pathStates = {
  I13N: ["data", "I13N"]
};

var myConnectOptions = _objectSpread(_objectSpread({}, connectOptions), {}, {
  shouldComponentUpdate: function shouldComponentUpdate() {
    var thisThemePath = pageStore.getState().get("themePath");

    if (globalStore.path && globalStore.path !== thisThemePath) {
      return false;
    } else {
      return true;
    }
  },
  initStates,
  pathStates,
  storeLocator
});

var ReshowReturn = getReturn({
  useConnect: useConnect(myConnectOptions),
  displayName: "ReshowReturn"
});
export { myConnectOptions as connectOptions, ReshowReturn as Return };
