// reshow-constant/build/cjs/src/index.js
"use strict";

exports.__esModule = true;
exports.UNDEFINED = exports.T_UNDEFINED = exports.T_TRUE = exports.T_NULL = exports.T_FALSE = exports.TYPE_ERROR = exports.SYMBOL = exports.STRING = exports.SCRIPT = exports.REAL_TIME_URL = exports.REAL_TIME_DATA_KEY = exports.OBJ_SIZE = exports.OBJECT = exports.NUMBER = exports.NEW_OBJ = exports.KEYS = exports.IS_ARRAY = exports.HAS = exports.FUNCTION = exports.DEFAULT = void 0;
// @ts-check
var DEFAULT = "default";
exports.DEFAULT = DEFAULT;
var FUNCTION = "function";
exports.FUNCTION = FUNCTION;
var NUMBER = "number";
exports.NUMBER = NUMBER;
var OBJECT = "object";
exports.OBJECT = OBJECT;
var STRING = "string";
exports.STRING = STRING;
var SYMBOL = "symbol";
exports.SYMBOL = SYMBOL;
var SCRIPT = "script";
exports.SCRIPT = SCRIPT;
var UNDEFINED = "undefined";
exports.UNDEFINED = UNDEFINED;
var TYPE_ERROR = "TypeError";
/**
 * @type undefined
 */

exports.TYPE_ERROR = TYPE_ERROR;
var T_UNDEFINED = void 0;
exports.T_UNDEFINED = T_UNDEFINED;
var T_NULL = null;
exports.T_NULL = T_NULL;
var T_TRUE = true;
exports.T_TRUE = T_TRUE;
var T_FALSE = false;
exports.T_FALSE = T_FALSE;
var KEYS = Object.keys;
exports.KEYS = KEYS;
var IS_ARRAY = Array.isArray;
/**
 * @param {object} o
 * @returns {number}
 */

exports.IS_ARRAY = IS_ARRAY;

var OBJ_SIZE = function OBJ_SIZE(o) {
  return o ? KEYS(o).length : 0;
};
/**
 * @returns {object}
 */


exports.OBJ_SIZE = OBJ_SIZE;

var NEW_OBJ = function NEW_OBJ() {
  return Object.create(null);
};
/**
 * @param {object} obj
 * @param {string} key
 * @returns {boolean}
 */


exports.NEW_OBJ = NEW_OBJ;

var HAS = function HAS(obj, key) {
  return !!(obj && Object.prototype.hasOwnProperty.call(obj, key));
}; // reshow specific


exports.HAS = HAS;
var REAL_TIME_URL = "--rtime-url--";
exports.REAL_TIME_URL = REAL_TIME_URL;
var REAL_TIME_DATA_KEY = "--rtime-data--";
exports.REAL_TIME_DATA_KEY = REAL_TIME_DATA_KEY;
