// organism-react-animate/build/es/src/aniUtil.mjs
//@ts-check
import { reactStyle, injectStyle } from "react-atomic-molecule";
import { NEW_OBJ } from "reshow-constant";
import getKeyframe from "keyframe-css";
import { initMap } from "get-object-value";
var inject = NEW_OBJ();
var injectDone = NEW_OBJ();
var injectCb = NEW_OBJ();
/**
 * @param {string} aniName
 */

var cleanTask = function cleanTask(aniName) {
  var i = injectCb[aniName].length;
  var tempArr = injectCb[aniName].splice(0, i);

  while (i--) {
    tempArr[i](injectDone[aniName]);
  }

  injectStyle();
};
/**
 * @param {string} key
 * @param {string} aniName
 * @param {number} timeout
 * @param {function} callback
 */


var initAni = function initAni(key, aniName, timeout, callback) {
  injectDone[aniName] ? callback() : initMap(injectCb)(aniName, []).push(callback);

  if (inject[key]) {
    return;
  }
  /**
   * @param {boolean} inject
   */


  var buildAniStyle = function buildAniStyle(inject) {
    reactStyle({
      animationName: [aniName],
      animationDuration: [timeout + 1 + "ms"],
      animationIterationCount: [1],
      animationTimingFunction: ["steps(" + Math.floor(timeout / 30) + ", end)"]
    }, "." + key, key);

    if (inject) {
      injectStyle();
    }
  };

  injectDone[aniName] ? buildAniStyle(true) : injectCb[aniName].push(buildAniStyle); // Need locate after reactStyle, for inject latest style in getKeyframe function

  getKeyframe(aniName, function () {
    injectDone[aniName] = true;
    cleanTask(aniName);
  });
  inject[key] = true;
};
/**
 * @typedef {object} AnimationData
 * @property {string} className
 * @property {string} key
 * @property {string} name
 * @property {number} timeout
 * @property {number} delay
 */

/**
 * @param {string} s
 * @returns {AnimationData}
 */


var parseAniValue = function parseAniValue(s) {
  if (s === void 0) {
    s = "";
  }

  var data = s.split("-");
  var name = data[0];
  var timeout = 500;
  var delay = 0;

  if (!isNaN(Number(data[1]))) {
    timeout = parseInt(data[1], 10);
  }

  if (!isNaN(Number(data[2]))) {
    delay = parseInt(data[2], 10);
    timeout += delay;
  }

  var key = [name, timeout, delay].join("-");
  return {
    className: key + " " + name,
    key,
    name,
    timeout,
    delay
  };
};

export { initAni, parseAniValue };
