// react-atomic-molecule/build/cjs/src/ui/molecules/Button.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _react = require("react");

var _classLib = require("class-lib");

var _SemanticUI = _interopRequireDefault(require("../molecules/SemanticUI.js"));

var _Icon = _interopRequireDefault(require("../molecules/Icon.js"));

var _useCSS = _interopRequireDefault(require("../../useCSS.js"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["type", "className", "children", "icon", "style"];

var Button = function Button(props) {
  (0, _useCSS["default"])(["button", "loader"], "semantic");
  var _props$type = props.type,
      type = _props$type === void 0 ? "button" : _props$type,
      className = props.className,
      children = props.children,
      icon = props.icon,
      style = props.style,
      others = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);
  var classes = (0, _classLib.mixClass)(className, "button");
  var thisIcon;
  var buttonWithIconStyle;

  if (icon) {
    thisIcon = /*#__PURE__*/(0, _jsxRuntime.jsx)(_Icon["default"], {
      style: Styles.icon,
      children: icon
    });
    buttonWithIconStyle = Styles.buttonWithIcon;
  }

  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_SemanticUI["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({
    atom: "button",
    type: type
  }, others), {}, {
    className: classes,
    style: (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, buttonWithIconStyle), style),
    children: [thisIcon, children]
  }));
};

var _default = Button;
exports["default"] = _default;
var Styles = {
  icon: {
    left: 20,
    top: 7,
    fill: "#fff",
    position: "absolute"
  },
  buttonWithIcon: {
    position: "relative",
    paddingLeft: 44
  }
};
module.exports = exports.default;
