// react-atomic-molecule/build/cjs/src/ui/molecules/Image.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _classLib = require("class-lib");

var _reshowBuild = _interopRequireDefault(require("reshow-build"));

var _SemanticUI = _interopRequireDefault(require("../molecules/SemanticUI.js"));

var _useCSS = _interopRequireDefault(require("../../useCSS.js"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["component", "loading", "src", "imgProps", "prepend"];

var Image = function Image(props) {
  (0, _useCSS["default"])(["image"], "semantic");
  var _props$component = props.component,
      component = _props$component === void 0 ? _SemanticUI["default"] : _props$component,
      _props$loading = props.loading,
      loading = _props$loading === void 0 ? "lazy" : _props$loading,
      src = props.src,
      imgProps = props.imgProps,
      prepend = props.prepend,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);
  var classes = (0, _classLib.mixClass)(props.className, {
    image: props.ui
  });
  var thisAlt = props.alt || props.title;
  var thisImgProps = (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, imgProps), {}, {
    atom: "img",
    alt: thisAlt,
    key: src,
    src: src,
    loading: loading
  });

  if (props.atom && "img" !== props.atom) {
    return (0, _reshowBuild["default"])(component)((0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, restProps), {}, {
      className: classes
    }), /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
      children: [prepend, /*#__PURE__*/(0, _jsxRuntime.jsx)(_SemanticUI["default"], (0, _objectSpread2["default"])({}, thisImgProps)), props.children]
    }));
  } else {
    return (0, _reshowBuild["default"])(component)((0, _objectSpread2["default"])((0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, restProps), thisImgProps), {}, {
      className: classes
    }));
  }
};

var _default = Image;
exports["default"] = _default;
module.exports = exports.default;
