// seturl/build/cjs/src/index.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

exports.__esModule = true;
exports.url = exports.unsetUrl = exports.parseUrl = exports.getUrlArray = exports.getUrl = exports["default"] = void 0;

var _winDoc = require("win-doc");

var _withArray = require("with-array");

var _reshowConstant = require("reshow-constant");

var _getKeyReg = _interopRequireWildcard(require("./getKeyReg.js"));

var _getUrlAnaly = _interopRequireDefault(require("./getUrlAnaly.js"));

var defaultValue = _reshowConstant.T_UNDEFINED;

var parseUrl = function parseUrl(url) {
  var oUrl = (0, _getUrlAnaly["default"])(url);
  return {
    host: oUrl[11],
    query: oUrl[16],
    path: oUrl[13]
  };
};

exports.parseUrl = parseUrl;

var resetUrl = function resetUrl(url) {
  return url ? url : (0, _winDoc.doc)().URL;
};

exports.url = resetUrl;

var getUrl = function getUrl(keys, origUrl) {
  var _parseUrl = parseUrl(resetUrl(origUrl)),
      _parseUrl$query = _parseUrl.query,
      query = _parseUrl$query === void 0 ? "" : _parseUrl$query;

  var getOne = function getOne(key) {
    var keyEq = key + "=";

    if (query.indexOf(keyEq) === query.lastIndexOf(keyEq)) {
      var reg = (0, _getKeyReg["default"])(key);
      var exec = reg.exec(query);
      return !exec ? defaultValue : decodeURIComponent(exec[3]);
    } else {
      var results = getUrlArray(key, query);
      return (0, _withArray.oneItemArrayToString)(results);
    }
  };

  if ((0, _reshowConstant.IS_ARRAY)(keys)) {
    var results = {};
    keys.forEach(function (key) {
      results[key] = getOne(key);
    });
    return results;
  } else {
    return keys ? getOne(keys) : defaultValue;
  }
};

exports.getUrl = getUrl;

var getMultiKey = function getMultiKey(key, query) {
  var reg = (0, _getKeyReg.getMultiMatchReg)(key);
  var results = [];
  var exec;

  while (exec = reg.exec(query)) {
    results.push(decodeURIComponent(exec[3]));
  }

  return results;
};

var getUrlArray = function getUrlArray(key, origUrl) {
  var _parseUrl2 = parseUrl(resetUrl(origUrl)),
      _parseUrl2$query = _parseUrl2.query,
      query = _parseUrl2$query === void 0 ? "" : _parseUrl2$query;

  return getMultiKey(key, query);
};

exports.getUrlArray = getUrlArray;

var unsetUrl = function unsetUrl(key, url) {
  url = resetUrl(url);
  var reg = (0, _getKeyReg["default"])(key);
  var exec;

  while (exec = reg.exec(url)) {
    url = exec[2] === "?" ? url.replace(reg, "?") : url.replace(reg, "");
  }

  return url;
};

exports.unsetUrl = unsetUrl;

var setUrl = function setUrl(key, value, url, KeepRawValue) {
  var multi = (0, _reshowConstant.IS_ARRAY)(value);
  url = unsetUrl(key, resetUrl(url));
  (multi ? value : [value]).forEach(function (vItem) {
    if (!KeepRawValue) {
      vItem = encodeURIComponent(vItem);
    }

    url = url + (-1 === url.indexOf("?") ? "?" : "&") + key + "=" + vItem;
  });
  return url;
};

var _default = setUrl;
exports["default"] = _default;
