// organism-react-animate/build/es/src/ui/organisms/AnimateImage.mjs
import _objectSpread from "reshow-runtime/es/helpers/objectSpread2";
import _objectWithoutProperties from "reshow-runtime/es/helpers/objectWithoutProperties";
var _excluded = ["src", "aniProps", "animate"];
import { useState, useEffect, useMemo } from "react";
import { Image } from "react-atomic-molecule";
import { win } from "win-doc";
import { useMounted } from "reshow-hooks";
import Change from "../organisms/Change.mjs";
import { jsx as _jsx } from "react/jsx-runtime";

var AnimateImage = function AnimateImage(props) {
  var {
    src,
    aniProps,
    animate = {
      enter: "fadeIn-300",
      leave: "fadeOut-300"
    }
  } = props,
      restProps = _objectWithoutProperties(props, _excluded);

  var [image, setImage] = useState(function () {
    return /*#__PURE__*/_jsx(Image, _objectSpread({
      src: src
    }, restProps));
  });

  var _mount = useMounted();

  useEffect(function () {
    var oImg = new (win().Image)();

    oImg.onload = function () {
      if (false !== _mount()) {
        setImage( /*#__PURE__*/_jsx(Image, _objectSpread({
          src: src
        }, restProps)));
      }
    };

    oImg.onerror = function () {
      console.warn("Get image failed. [" + src + "]");

      if (false !== _mount()) {
        setImage(null);
      }
    };

    oImg.src = src;
  }, [src]);
  return useMemo(function () {
    return /*#__PURE__*/_jsx(Change, _objectSpread(_objectSpread(_objectSpread({
      className: "animate-image"
    }, animate), aniProps), {}, {
      children: image
    }));
  }, [image, animate]);
};

export default AnimateImage;
