// with-array/build/cjs/src/index.js
"use strict";

exports.__esModule = true;
exports.testString = exports.stringToArray = exports.oneItemArrayToString = exports.objectToArray = exports["default"] = exports.arrayToObject = void 0;

var _reshowConstant = require("reshow-constant");

var objectToArray = function objectToArray(obj) {
  return (0, _reshowConstant.KEYS)(obj || {}).map(function (key) {
    return obj[key];
  });
};

exports.objectToArray = objectToArray;

var arrayToObject = function arrayToObject(arr, key) {
  var map = {};

  if ((0, _reshowConstant.IS_ARRAY)(arr)) {
    var i = arr.length;

    while (i--) {
      var a = arr[i];
      map[a[key]] = a;
    }
  }

  return map;
};

exports.arrayToObject = arrayToObject;

var oneItemArrayToString = function oneItemArrayToString(maybeArray) {
  if ((0, _reshowConstant.IS_ARRAY)(maybeArray)) {
    return maybeArray.length > 1 ? maybeArray : maybeArray[0];
  } else {
    return maybeArray;
  }
}; // contain null, undefined, numeric, string


exports.oneItemArrayToString = oneItemArrayToString;

var testString = function testString(s) {
  return _reshowConstant.T_NULL == s || !isNaN(parseInt(s)) || _reshowConstant.STRING === typeof s;
};

exports.testString = testString;

var stringToArray = function stringToArray(maybeString) {
  return anyToArray(maybeString, testString);
};

exports.stringToArray = stringToArray;

var anyToArray = function anyToArray(val, testFunc) {
  testFunc = testFunc || function (v) {
    return !(0, _reshowConstant.IS_ARRAY)(v);
  };

  return testFunc(val) ? val === _reshowConstant.T_UNDEFINED ? [] : [val] : val;
};

var _default = anyToArray;
exports["default"] = _default;
