// reshow-url/build/cjs/src/ui/organisms/ClientRoute.js
"use strict";

var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("reshow-runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("reshow-runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("reshow-runtime/helpers/get"));

var _getPrototypeOf2 = _interopRequireDefault(require("reshow-runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("reshow-runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("reshow-runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("reshow-runtime/helpers/defineProperty"));

var _reshow = _interopRequireWildcard(require("reshow"));

var _organismReactAjax = require("organism-react-ajax");

var _winDoc = require("win-doc");

var _handleAnchor = _interopRequireDefault(require("../../handleAnchor.js"));

//@ts-check

/**
 * @param {string} url
 */
var defaultOnUrlChange = function defaultOnUrlChange(url) {
  return (
    /**
     * @param {function} handleAnchor
     */
    function (handleAnchor) {
      return (
        /**
         * @param {number} goAnchorDelay
         * @returns {Object}
         */
        function (goAnchorDelay) {
          var separator = "/";
          var params = url.split(separator);
          var last = params.length - 1;
          var lastPath = params[last];
          var next = {
            pvid: url,
            themePath: null
          };

          if (lastPath) {
            next.themePath = handleAnchor(lastPath)(goAnchorDelay);
          }

          return next;
        }
      );
    }
  );
};
/**
 * @extends {Reshow}
 */


var ClientRoute = /*#__PURE__*/function (_Reshow) {
  (0, _inherits2["default"])(ClientRoute, _Reshow);

  var _super = (0, _createSuper2["default"])(ClientRoute);

  function ClientRoute() {
    (0, _classCallCheck2["default"])(this, ClientRoute);
    return _super.apply(this, arguments);
  }

  (0, _createClass2["default"])(ClientRoute, [{
    key: "getPath",
    value: function getPath() {
      var _this$getUrlChangeSta,
          _this = this;

      var themePath = (_this$getUrlChangeSta = this.getUrlChangeState(this.props.url || (0, _winDoc.doc)().URL)) === null || _this$getUrlChangeSta === void 0 ? void 0 : _this$getUrlChangeSta.themePath;
      var themes = this.props.themes;

      if (themes[themePath]) {
        setTimeout(function () {
          return (0, _reshow.dispatch)({
            themePath: themePath !== null && themePath !== void 0 ? themePath : _this.getGlobalPath()
          });
        });
        return themePath;
      }
    }
    /**
     * @param {string} url
     * @returns {Object}
     */

  }, {
    key: "getUrlChangeState",
    value: function getUrlChangeState(url) {
      var _this$props = this.props,
          onUrlChange = _this$props.onUrlChange,
          onHashChange = _this$props.onHashChange,
          goAnchorDelay = _this$props.goAnchorDelay;
      var thisUrlChangeFunc = onUrlChange ? onUrlChange : defaultOnUrlChange;
      return thisUrlChangeFunc(url)(onHashChange)(goAnchorDelay);
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      (0, _get2["default"])((0, _getPrototypeOf2["default"])(ClientRoute.prototype), "componentDidMount", this).call(this);
      (0, _organismReactAjax.ajaxDispatch)({
        onUrlChange: this.getUrlChangeState.bind(this)
      });
    }
  }]);
  return ClientRoute;
}(_reshow["default"]);

(0, _defineProperty2["default"])(ClientRoute, "defaultProps", {
  ajax: false,
  goAnchorDelay: 1500,
  fallback: "div",
  onHashChange: _handleAnchor["default"]
});
var _default = ClientRoute;
exports["default"] = _default;
module.exports = exports.default;
