// reshow/build/cjs/src/ui/molecules/ReshowComponent.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.connectOptions = exports.Return = void 0;

var _objectSpread2 = _interopRequireDefault(
  require("reshow-runtime/helpers/objectSpread2")
);

var _reshowBuild = _interopRequireDefault(require("reshow-build"));

var _reshowReturn = require("reshow-return");

var _reshowFlux = require("reshow-flux");

var _pageStore = _interopRequireDefault(require("../../stores/pageStore.js"));

var _globalStore = require("../../stores/globalStore.js");

var storeLocator = function storeLocator(props) {
  return (
    (props === null || props === void 0 ? void 0 : props.store) ||
    _pageStore["default"]
  );
};

var initStates = ["data", "I18N"];
var pathStates = {
  I13N: ["data", "I13N"],
};
var myConnectOptions = (0, _objectSpread2["default"])(
  (0, _objectSpread2["default"])({}, _reshowReturn.connectOptions),
  {},
  {
    shouldComponentUpdate: function shouldComponentUpdate() {
      var thisThemePath = _pageStore["default"].getState().get("themePath");

      if (
        _globalStore.globalStore.path &&
        _globalStore.globalStore.path !== thisThemePath
      ) {
        return false;
      } else {
        return true;
      }
    },
    initStates: initStates,
    pathStates: pathStates,
    storeLocator: storeLocator,
  }
);
exports.connectOptions = myConnectOptions;
var ReshowReturn = (0, _reshowReturn.getReturn)({
  useConnect: (0, _reshowFlux.useConnect)(myConnectOptions),
  displayName: "ReshowReturn",
});
exports.Return = ReshowReturn;
