// reshow/build/cjs/src/ui/organisms/ReshowMessage.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(
  require("reshow-runtime/helpers/objectSpread2")
);

var _react = require("react");

var _organismReactPopup = require("organism-react-popup");

var _reactAtomicMolecule = require("react-atomic-molecule");

var _reshowFlux = require("reshow-flux");

var _getObjectValue = _interopRequireDefault(require("get-object-value"));

var _ReshowComponent = require("../molecules/ReshowComponent.js");

var _messageStore = _interopRequireDefault(
  require("../../stores/messageStore.js")
);

var _index = require("../../index.js");

var _jsxRuntime = require("react/jsx-runtime");

// Local import
var handleDismiss = function handleDismiss(e) {
  var _e$data;

  var id =
    e === null || e === void 0
      ? void 0
      : (_e$data = e.data) === null || _e$data === void 0
      ? void 0
      : _e$data.id;

  if (id) {
    (0, _index.dispatch)("alert/del", {
      id: id,
    });
  }
};

var handleClick = function handleClick(dialog) {
  return function (e) {
    var value = (0, _getObjectValue["default"])(e, ["currentTarget", "value"]);
    setTimeout(function () {
      if (dialog) {
        (0, _index.dispatch)("dialog/end", {
          value: value,
        });
      }
    });
  };
};

var Body = function Body(props) {
  var defaultAlertProps = props.defaultAlertProps,
    defaultDialogProps = props.defaultDialogProps,
    alerts = props.alerts,
    alertProps = props.alertProps,
    _props$alertComponent = props.alertComponent,
    alertComponent =
      _props$alertComponent === void 0
        ? _organismReactPopup.AlertsNotifier
        : _props$alertComponent,
    dialog = props.dialog,
    dialogProps = props.dialogProps,
    _props$dialogComponen = props.dialogComponent,
    dialogComponent =
      _props$dialogComponen === void 0
        ? _organismReactPopup.Dialog
        : _props$dialogComponen;
  var thisDialog = null;

  if (dialog) {
    thisDialog = /*#__PURE__*/ (0, _jsxRuntime.jsx)(
      _organismReactPopup.DisplayPopupEl,
      {
        children: (0, _reactAtomicMolecule.build)(dialogComponent)(
          (0, _objectSpread2["default"])(
            (0, _objectSpread2["default"])(
              (0, _objectSpread2["default"])({}, defaultDialogProps),
              (0, _reshowFlux.toJS)(dialogProps)
            ),
            {},
            {
              onClick: handleClick(dialog),
              name: "reshow-dialog",
            }
          ),
          (0, _reshowFlux.toJS)(dialog)
        ),
      }
    );
  }

  return /*#__PURE__*/ (0, _jsxRuntime.jsxs)(_reactAtomicMolecule.SemanticUI, {
    children: [
      thisDialog,
      (0, _reactAtomicMolecule.build)(alertComponent)(
        (0, _objectSpread2["default"])(
          (0, _objectSpread2["default"])(
            (0, _objectSpread2["default"])({}, defaultAlertProps),
            (0, _reshowFlux.toJS)(alertProps)
          ),
          {},
          {
            onDismiss: handleDismiss,
            alerts: (0, _reshowFlux.toJS)(alerts),
            name: "reshow-alerts",
          }
        )
      ),
    ],
  });
};

var ReshowMessage = /*#__PURE__*/ (0, _react.memo)(function (props) {
  return /*#__PURE__*/ (0, _jsxRuntime.jsx)(_ReshowComponent.Return, {
    store: _messageStore["default"],
    initStates: ["alerts", "alertProps", "dialog", "dialogProps"],
    children: /*#__PURE__*/ (0, _jsxRuntime.jsx)(
      Body,
      (0, _objectSpread2["default"])({}, props)
    ),
  });
});
ReshowMessage.displayName = "ReshowMessage";
var _default = ReshowMessage;
exports["default"] = _default;
module.exports = exports.default;
