// reshow/build/es/src/stores/clientStorageStore.mjs
import { createReducer } from "reshow-flux-base";
import { localStorage, sessionStorage, Storage } from "get-storage";
import { win } from "win-doc";
var isUpdating;
var lastDispatch = {
  current: {}
};

var getLocalReducer = function getLocalReducer(name, storage, disableEncode) {
  return createReducer(function (state, action) {
    isUpdating = true;
    lastDispatch.current = {
      name,
      action
    };
    var result = state.merge(action);
    isUpdating = false;
    return result;
  }, new Storage(storage, disableEncode));
};

var LOCAL_STORAGE = "localStorage";
var LOCAL_VALUE = "localValue";
var SESSION_STORAGE = "sessionStorage";
var SESSION_VALUE = "sessionValue";
var [localStorageStore, localStorageDispatch] = getLocalReducer(LOCAL_STORAGE, localStorage);
var [localValueStore, localValueDispatch] = getLocalReducer(LOCAL_VALUE, localStorage, true);
var [sessionStorageStore, sessionStorageDispatch] = getLocalReducer(SESSION_STORAGE, sessionStorage);
var [sessionValueStore, sessionValueDispatch] = getLocalReducer(SESSION_VALUE, sessionStorage, true);
export { localStorageStore, localStorageDispatch, localValueStore, localValueDispatch, sessionStorageStore, sessionStorageDispatch, sessionValueStore, sessionValueDispatch };
