// react-atomic-molecule/build/cjs/src/ui/molecules/Form.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _classLib = require("class-lib");

var _SemanticUI = _interopRequireDefault(require("../molecules/SemanticUI.js"));

var _useCSS = _interopRequireDefault(require("../../useCSS.js"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["messageType", "style", "className"];

var Form = function Form(props) {
  (0, _useCSS["default"])(["form", "input", "loader", "search"], "semantic");
  var messageType = props.messageType,
      style = props.style,
      className = props.className,
      others = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);
  var classes = (0, _classLib.mixClass)(className, messageType, "form");
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_SemanticUI["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({
    atom: "form"
  }, others), {}, {
    className: classes,
    style: (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, Styles.form), style)
  }));
};

var _default = Form;
exports["default"] = _default;
var Styles = {
  form: {
    boxSizing: "border-box"
  }
};
module.exports = exports.default;
