// organism-react-popup/build/es/src/ui/molecules/BasePopup.mjs
import _classCallCheck from "reshow-runtime/es/helpers/classCallCheck";
import _createClass from "reshow-runtime/es/helpers/createClass";
import _assertThisInitialized from "reshow-runtime/es/helpers/assertThisInitialized";
import _inherits from "reshow-runtime/es/helpers/inherits";
import _createSuper from "reshow-runtime/es/helpers/createSuper";
import _defineProperty from "reshow-runtime/es/helpers/defineProperty";
import React, { PureComponent } from "react";
import needCSS from "need-css";
import { popupDispatch } from "../../stores/popupStore.mjs";

var BasePopup = /*#__PURE__*/function (_PureComponent) {
  _inherits(BasePopup, _PureComponent);

  var _super = _createSuper(BasePopup);

  function BasePopup(props) {
    var _this;

    _classCallCheck(this, BasePopup);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "state", {
      hasError: false
    });

    needCSS(["popup", "modal"], "semantic");
    return _this;
  }

  _createClass(BasePopup, [{
    key: "componentDidCatch",
    value: function componentDidCatch(error, info) {
      var {
        onError
      } = this.props;

      if (onError) {
        callfunc(onError, [error, info]);
      } else {
        console.error([error, info]);
      }

      this.setState({
        hasError: true
      });
    }
  }, {
    key: "close",
    value: function close() {
      popupDispatch({
        type: "dom/closeOne",
        params: {
          popup: this
        }
      });
      return true;
    }
  }], [{
    key: "getDerivedStateFromError",
    value: function getDerivedStateFromError(error) {
      return {
        hasError: true
      };
    }
  }]);

  return BasePopup;
}(PureComponent);

_defineProperty(BasePopup, "defaultProps", {
  name: "default"
});

export default BasePopup;
