// organism-react-ajax/build/cjs/src/worker.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");
var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");
exports.__esModule = true;
exports["default"] = void 0;
var _classCallCheck2 = _interopRequireDefault(require("reshow-runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("reshow-runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("reshow-runtime/helpers/defineProperty"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));
var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));
var _getObjectValue = _interopRequireWildcard(require("get-object-value"));
var _nonWorker = _interopRequireDefault(require("non-worker"));
var _superagent = _interopRequireDefault(require("superagent"));
var _parseIniString = _interopRequireDefault(require("parse-ini-string"));
var _objectNested = require("object-nested");
var _reshowConstant = require("reshow-constant");
var _excluded = ["error", "req", "text", "xhr"],
  _excluded2 = ["id", "query", "isSendJson", "cookHeaders", "responseType"],
  _excluded3 = ["error", "req", "text", "xhr"];
var _this = void 0;
var arrWs = {};
var arrReq = {};
var getJson = function getJson(text) {
  var json;
  try {
    json = JSON.parse(text);
  } catch (e) {}
  return json || {};
};
var oNonWorker = new _nonWorker["default"](function (e) {
  var data = (0, _getObjectValue["default"])(e, ["data"]);
  switch (data.type) {
    case "initWs":
      initWs(data.ws)(data.params);
      break;
    case "closeWs":
      closeWs(data.ws);
      break;
    case "ajaxGet":
      ajaxGet(data);
      break;
    case "ajaxPost":
      ajaxPost(data);
      break;
  }
});
var post = function post(payload) {
  var strWcb = (0, _getObjectValue["default"])(payload, ["params", "workerCallback"]);
  var parseIni = (0, _getObjectValue["default"])(payload, ["params", "ini"]);
  var text = (0, _getObjectValue["default"])(payload, ["params", "text"]);
  payload.params.json = parseIni ? (0, _objectNested.nest)((0, _parseIniString["default"])(text), "_") : getJson(text);
  if (strWcb) {
    var wcb = eval("(" + strWcb + ")");
    payload = wcb(payload);
  }
  oNonWorker.post.call(_this, payload);
};
var _default = oNonWorker;
exports["default"] = _default;
var cookParams = function cookParams(action, callReq) {
  var params = (0, _getObjectValue["default"])(action, ["params"], {});
  var id = params.id;
  if (id) {
    if (arrReq[id]) {
      arrReq[id].abort();
    }
    arrReq[id] = callReq;
  }
  var cookHeaders = (0, _objectSpread2["default"])((0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, (0, _getObjectValue["default"])(params, ["globalHeaders"], {})), (0, _getObjectValue["default"])(params, ["headers"], {})), {}, {
    Accept: (0, _getObjectValue["default"])(params, ["accept"], "application/json")
  });
  params.cookHeaders = cookHeaders;
  var superagent = params.superagent || {};
  var syncKeys = ["responseType"];
  syncKeys.forEach(function (key) {
    if (params[key]) {
      superagent[key] = params[key];
    }
  });
  (0, _reshowConstant.KEYS)(superagent).forEach(function (key) {
    callReq = callReq[key].apply(callReq, superagent[key]);
  });
  return params;
};
var ajaxGet = function ajaxGet(_ref) {
  var url = _ref.url,
    action = _ref.action;
  var callReq = _superagent["default"].get(url);
  var _cookParams = cookParams(action, callReq),
    query = _cookParams.query,
    cookHeaders = _cookParams.cookHeaders,
    id = _cookParams.id;
  callReq.query(query).set(cookHeaders).end(function (err, res) {
    if (res) {
      if (arrReq[id]) {
        delete arrWs[id];
      }
      var error = res.error,
        _req = res.req,
        text = res.text,
        xhr = res.xhr,
        response = (0, _objectWithoutPropertiesLoose2["default"])(res, _excluded);
      action.params = (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, action.params), {}, {
        text: text,
        response: response
      });
      post(action);
    }
  });
};
var ajaxPost = function ajaxPost(_ref2) {
  var url = _ref2.url,
    action = _ref2.action;
  var callReq;
  switch ((0, _getObjectValue["default"])(action, ["params", "method"])) {
    case "delete":
      callReq = _superagent["default"].del(url);
      break;
    case "head":
      callReq = _superagent["default"].head(url);
      break;
    case "patch":
      callReq = _superagent["default"].patch(url);
      break;
    case "put":
      callReq = _superagent["default"].put(url);
      break;
    default:
      callReq = _superagent["default"].post(url);
      break;
  }
  var _cookParams2 = cookParams(action, callReq),
    id = _cookParams2.id,
    query = _cookParams2.query,
    isSendJson = _cookParams2.isSendJson,
    cookHeaders = _cookParams2.cookHeaders,
    responseType = _cookParams2.responseType,
    params = (0, _objectWithoutPropertiesLoose2["default"])(_cookParams2, _excluded2);
  var isSend = false;
  if (isSendJson) {
    isSend = true;
  } else {
    if (null == isSendJson && query) {
      (0, _reshowConstant.KEYS)(query).every(function (key) {
        if ("object" !== typeof query[key]) {
          return true;
        }
        isSend = true;
        return false;
      });
    }
  }
  if (!isSend) {
    callReq = callReq.type("form");
  }
  callReq.send(query).set(cookHeaders).end(function (err, res) {
    if (res) {
      if (arrReq[id]) {
        delete arrWs[id];
      }
      var error = res.error,
        _req2 = res.req,
        text = res.text,
        xhr = res.xhr,
        response = (0, _objectWithoutPropertiesLoose2["default"])(res, _excluded3);
      action.params = (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, action.params), {}, {
        text: text,
        response: response
      });
      post(action);
    }
  });
};
var closeWs = function closeWs(url) {
  if (arrWs[url]) {
    arrWs[url].close();
    delete arrWs[url];
  }
  return !arrWs[url];
};
var WebSocketHelper = /*#__PURE__*/function () {
  function WebSocketHelper(url, params) {
    (0, _classCallCheck2["default"])(this, WebSocketHelper);
    (0, _defineProperty2["default"])(this, "isWsConnect", false);
    this.ws = null;
    this.pingTimeout = null;
    this.url = url;
    this.params = params;
    this.open();
  }
  (0, _createClass2["default"])(WebSocketHelper, [{
    key: "open",
    value: function open() {
      var _this2 = this;
      if (this.isWsConnect) {
        return;
      }
      var url = this.url;
      var params = this.params;

      /**
       * Can't try catch
       * WebSocket connection to 'xxx' failed
       * use onerror instead.
       */
      var ws = new WebSocket(url);
      this.ws = ws;
      ws.onopen = function (e) {
        _this2.isWsConnect = true;
        _this2.ping();
        var messages = params.messages;
        if ((0, _getObjectValue["default"])(messages, ["length"])) {
          messages.forEach(function (m) {
            return ws.send(JSON.stringify(m));
          });
        }
      };
      ws.onerror = function (e) {
        _this2.isWsConnect = false;
        _this2.ping();
      };
      ws.onmessage = function (e) {
        switch (e.data) {
          case "pong":
            break;
          default:
            post({
              type: "ws",
              params: {
                text: e.data,
                url: url
              }
            });
            break;
        }
      };
      ws.onclose = function (e) {
        _this2.isWsConnect = false;
        console.warn("WS close.", url);
      };
    }
  }, {
    key: "close",
    value: function close() {
      this.ws.close();
      this.clearPing();
    }
  }, {
    key: "clearPing",
    value: function clearPing() {
      if (this.pingTimeout) {
        clearTimeout(this.pingTimeout);
        this.pingTimeout = null;
      }
    }
  }, {
    key: "ping",
    value: function ping() {
      var _this3 = this;
      this.clearPing();
      this.pingTimeout = setTimeout(function () {
        if (!_this3.isWsConnect) {
          console.warn("Try restore ws connection.", _this3.url);
          _this3.open();
        } else {
          _this3.ws.send(JSON.stringify({
            type: "ping"
          }));
        }
        _this3.ping();
      }, 15000);
    }
  }]);
  return WebSocketHelper;
}();
var initWs = function initWs(url) {
  return function (params) {
    var create = function create() {
      arrWs[url] = new WebSocketHelper(url, params);
    };
    if (!arrWs[url]) {
      create(url);
    }
  };
};
module.exports = exports.default;
