// reshow-constant/build/es/src/index.mjs
// @ts-check
export var DEFAULT = "default";
export var FUNCTION = "function";
export var NUMBER = "number";
export var OBJECT = "object";
export var STRING = "string";
export var SYMBOL = "symbol";
export var SCRIPT = "script";
export var UNDEFINED = "undefined";
export var TYPE_ERROR = "TypeError";
/**
 * @type undefined
 */

export var T_UNDEFINED = void 0;
export var T_NULL = null;
export var T_TRUE = true;
export var T_FALSE = false;
export var KEYS = Object.keys;
export var IS_ARRAY = Array.isArray;
/**
 * @param {object} o
 * @returns {number}
 */

export var OBJ_SIZE = function OBJ_SIZE(o) {
  return o ? KEYS(o).length : 0;
};
/**
 * @returns {object}
 */

export var NEW_OBJ = function NEW_OBJ() {
  return Object.create(null);
};
/**
 * @param {object} obj
 * @param {string} key
 * @returns {boolean}
 */

export var HAS = function HAS(obj, key) {
  return !!(obj && Object.prototype.hasOwnProperty.call(obj, key));
}; // reshow specific

export var REAL_TIME_URL = "--rtime-url--";
export var REAL_TIME_DATA_KEY = "--rtime-data--";
