// get-scroll-info/build/es/src/index.mjs
// @ts-check

import { doc, win } from "win-doc";
import { UNDEFINED, NEW_OBJ } from "reshow-constant";
var lastScrollStore = NEW_OBJ();
/**
 * @typedef {object & Document} ThisDocument
 * @propert {function} webkitIsFullScreen
 */
/**
 * @type ThisDocument
 */
var oDoc;
var oWin;
var isWebkit;
var docEl;
var domCount = 0;
var initDoc = function initDoc() {
  oDoc = doc();
  oWin = win();
  isWebkit = UNDEFINED !== typeof oDoc.webkitIsFullScreen;
  docEl = oDoc.documentElement;
};

/**
 * @param {HTMLElement} el
 * @returns {HTMLElement}
 */
var getScrollNode = function getScrollNode(el) {
  if (!oDoc) {
    initDoc();
  }
  if (!el && oDoc) {
    if (oDoc.scrollingElement) {
      el = oDoc.scrollingElement;
    } else if (isWebkit) {
      el = oDoc.body;
    } else {
      el = docEl;
    }
  }
  if (!el.id) {
    el.id = "scroll-info-" + domCount;
    domCount++;
  }
  return el;
};

/**
 * @typedef {object} InfoType 
 * @property {boolean} atTop
 * @property {boolean} atRight
 * @property {boolean} atBottom 
 * @property {boolean} atLeft 
 * @property {boolean} isScrollUp 
 * @property {boolean} isScrollRight 
 * @property {boolean} isScrollDown 
 * @property {boolean} isScrollLeft
 * @property {number} scrollWidth 
 * @property {number} scrollHeight 
 * @property {number} scrollNodeWidth 
 * @property {number} scrollNodeHeight 
 * @property {number} top
 * @property {number} right 
 * @property {number} bottom 
 * @property {number} left 
 */

/**
 * @returns {InfoType}
 */
var getScrollInfo = function getScrollInfo(el, margin) {
  if (el === void 0) {
    el = null;
  }
  if (margin === void 0) {
    margin = 50;
  }
  el = getScrollNode(el);
  var w;
  var h;
  var nodeName = (el.nodeName || "").toLowerCase();
  var isRoot = "body" === nodeName || "html" === nodeName;

  // defined scrollWidth and scrollHeight first.
  // to know if have vertical or horizontal bar.
  var scrollWidth = el.scrollWidth;
  var scrollHeight = el.scrollHeight;
  if (isRoot) {
    w = Math.max(el.clientWidth || 0, oWin.innerWidth || 0);
    h = Math.max(el.clientHeight || 0, oWin.innerHeight || 0);
    var hasHorizontalBar = w < scrollWidth;
    var hasVerticalBar = h < scrollHeight;
    if (hasHorizontalBar) {
      h = el.clientHeight;
    }
    if (hasVerticalBar) {
      w = el.clientWidth;
    }
  } else {
    w = el.clientWidth;
    h = el.clientHeight;
  }
  var scrollLeft = el.scrollLeft;
  var scrollTop = el.scrollTop;
  var scrollBottom = scrollTop + h;
  var scrollRight = scrollLeft + w;
  var elId = el.id;
  var lastScroll = lastScrollStore[elId];
  var info = {
    atTop: scrollTop < margin,
    atRight: scrollRight > scrollWidth - margin,
    atBottom: scrollBottom > scrollHeight - margin,
    atLeft: scrollLeft < margin,
    isScrollUp: lastScroll && scrollTop < lastScroll.top,
    isScrollRight: lastScroll && scrollLeft > lastScroll.left,
    isScrollDown: lastScroll && scrollTop > lastScroll.top,
    isScrollLeft: lastScroll && scrollLeft < lastScroll.left,
    scrollWidth,
    scrollHeight,
    scrollNodeWidth: w,
    scrollNodeHeight: h,
    top: scrollTop,
    right: scrollRight,
    bottom: scrollBottom,
    left: scrollLeft
  };
  lastScrollStore[elId] = info;
  return info;
};
export default getScrollInfo;
export { getScrollNode };
