// reshow-url/build/cjs/src/handleAnchor.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.goToAnchor = exports.getAnchorPath = exports.disableHandleAnchor = exports["default"] = void 0;

var _getoffset = _interopRequireDefault(require("getoffset"));

var _smoothScrollTo = _interopRequireDefault(require("smooth-scroll-to"));

var _winDoc = require("win-doc");

var goAnchorTimer;

var urlDecode = function urlDecode(s) {
  return decodeURIComponent(s);
};

var goToAnchor = function goToAnchor(anchor) {
  return function (goAnchorDelay) {
    if (!goAnchorDelay) {
      goAnchorDelay = 0;
    }

    clearTimeout(goAnchorTimer);
    goAnchorTimer = setTimeout(function () {
      try {
        var dom = document.body.querySelector(anchor);

        if (dom) {
          var pos = (0, _getoffset["default"])(dom);
          (0, _smoothScrollTo["default"])(pos.top);
        }
      } catch (e) {}
    }, goAnchorDelay);
  };
};

exports.goToAnchor = goToAnchor;

var getAnchorPath = function getAnchorPath(path) {
  if (!path) {
    path = (0, _winDoc.doc)().URL;
  } else {
    if (0 === path.indexOf("#")) {
      path = "/" + path;
    }
  }

  var pathArr = path.split("/#/");

  if (null != pathArr[1]) {
    path = "/" + pathArr[1];
  }

  var anchor = "";
  var hashStart = path.indexOf("#");
  var anchorStart = -1 !== hashStart ? hashStart : path.indexOf("%23");

  if (-1 !== anchorStart) {
    anchor = urlDecode(path.substring(anchorStart));
    path = path.substring(0, anchorStart);
  }

  var anchorArr = anchor.split("#");
  return {
    anchor: anchor,
    path: path,
    anchorArr: anchorArr,
    lastAnchor: "#" + anchorArr[anchorArr.length - 1]
  };
};

exports.getAnchorPath = getAnchorPath;

var handleAnchor = function handleAnchor(rawPath) {
  return function (goAnchorDelay) {
    var _getAnchorPath = getAnchorPath(rawPath),
        anchor = _getAnchorPath.anchor,
        path = _getAnchorPath.path;

    if (anchor) {
      goToAnchor(anchor)(goAnchorDelay);
    }

    if (-1 !== path.indexOf("?")) {
      return path.split("?")[0];
    } else {
      return path;
    }
  };
};

var disableHandleAnchor = function disableHandleAnchor(path) {
  return function () {
    return path;
  };
};

exports.disableHandleAnchor = disableHandleAnchor;
var _default = handleAnchor;
exports["default"] = _default;
