// with-array/build/es/src/index.mjs
import { IS_ARRAY, T_NULL, T_UNDEFINED, KEYS, STRING } from "reshow-constant";

var objectToArray = function objectToArray(obj) {
  return KEYS(obj || {}).map(function (key) {
    return obj[key];
  });
};

var arrayToObject = function arrayToObject(arr, key) {
  var map = {};

  if (IS_ARRAY(arr)) {
    var i = arr.length;

    while (i--) {
      var a = arr[i];
      map[a[key]] = a;
    }
  }

  return map;
};

var oneItemArrayToString = function oneItemArrayToString(maybeArray) {
  if (IS_ARRAY(maybeArray)) {
    return maybeArray.length > 1 ? maybeArray : maybeArray[0];
  } else {
    return maybeArray;
  }
}; // contain null, undefined, numeric, string


var testString = function testString(s) {
  return T_NULL == s || !isNaN(parseInt(s)) || STRING === typeof s;
};

var stringToArray = function stringToArray(maybeString) {
  return anyToArray(maybeString, testString);
};

var anyToArray = function anyToArray(val, testFunc) {
  testFunc = testFunc || function (v) {
    return !IS_ARRAY(v);
  };

  return testFunc(val) ? val === T_UNDEFINED ? [] : [val] : val;
};

export default anyToArray;
export { oneItemArrayToString, objectToArray, arrayToObject, testString, stringToArray };
