// reshow/build/cjs/src/ui/organisms/RealTimeReturn.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(
  require("reshow-runtime/helpers/objectSpread2")
);

var _getObjectValue = _interopRequireDefault(require("get-object-value"));

var _reshowFlux = require("reshow-flux");

var _reactAtomicMolecule = require("react-atomic-molecule");

var _reshowReturn = require("reshow-return");

var _reshowConstant = require("reshow-constant");

var _ReshowComponent = require("../molecules/ReshowComponent.js");

var _realTimeStore = _interopRequireDefault(
  require("../../stores/realTimeStore.js")
);

var calculateState = function calculateState(prevState, options) {
  /**
   * storeState was pass from reducer directly to avoid synchronous get wrong data.
   */
  var path = options.realTimePath,
    url = options.realTimeUrl,
    realTimeReset = options.realTimeReset,
    realTimeState = options.storeSyncState;

  if ((0, _reshowConstant.IS_ARRAY)(path) && path.length) {
    path.unshift(_reshowConstant.REAL_TIME_DATA_KEY);
  }

  var state = (0, _getObjectValue["default"])(
    realTimeState,
    path || [_reshowConstant.REAL_TIME_DATA_KEY]
  );
  var wsUrl = (0, _getObjectValue["default"])(realTimeState, [
    _reshowConstant.REAL_TIME_URL,
  ]);

  if (state && (!url || url === wsUrl)) {
    state[_reshowConstant.REAL_TIME_URL] = wsUrl;
    return state;
  } else {
    if (realTimeReset) {
      // Reset for when reconnection to new websocket server
      // will not send duplicate data to client
      var reset = {};
      (0, _reshowConstant.KEYS)(prevState).forEach(function (key) {
        return (reset[key] = null);
      });
      return reset;
    } else {
      return prevState;
    }
  }
};

var storeLocator = function storeLocator(props) {
  return props.store || _realTimeStore["default"];
};

var myConnectOptions = (0, _objectSpread2["default"])(
  (0, _objectSpread2["default"])({}, _ReshowComponent.connectOptions),
  {},
  {
    calculateState: calculateState,
    realTimePath: null,
    realTimeUrl: null,
    realTimeReset: false,
    storeLocator: storeLocator,
  }
);
var RealTimeReturn = (0, _reshowReturn.getReturn)({
  cleanProps: ["realTimePath", "realTimeUrl", "realTimeReset"],
  useConnect: (0, _reshowFlux.useConnect)(myConnectOptions),
  displayName: "RealTimeReturn",
});
var _default = RealTimeReturn;
exports["default"] = _default;
module.exports = exports.default;
