// react-atomic-molecule/build/es/src/ui/molecules/Message.mjs
import _objectSpread from "reshow-runtime/es/helpers/objectSpread2";
import _objectWithoutProperties from "reshow-runtime/es/helpers/objectWithoutProperties";
var _excluded = ["header", "className", "children", "messageType"];
import { mixClass } from "class-lib";
import Header from "../molecules/Header.mjs";
import SemanticUI from "../molecules/SemanticUI.mjs";
import useCSS from "../../useCSS.mjs";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";

var Message = function Message(props) {
  useCSS(["message"], "semantic");

  var {
    header,
    className,
    children,
    messageType
  } = props,
      reset = _objectWithoutProperties(props, _excluded);

  var classes = mixClass(className, {
    info: messageType === "info",
    error: messageType === "error",
    success: messageType === "success",
    warning: messageType === "warning"
  }, "message");
  var thisHeader;

  if (header) {
    thisHeader = /*#__PURE__*/_jsx(Header, {
      children: header
    });
  }

  return /*#__PURE__*/_jsxs(SemanticUI, _objectSpread(_objectSpread({}, reset), {}, {
    className: classes,
    children: [thisHeader, children]
  }));
};

export default Message;
