// organism-react-animate/build/cjs/src/ui/organisms/AnimateImage.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _react = require("react");

var _reactAtomicMolecule = require("react-atomic-molecule");

var _winDoc = require("win-doc");

var _reshowHooks = require("reshow-hooks");

var _Change = _interopRequireDefault(require("../organisms/Change.js"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["src", "aniProps", "animate"];

var AnimateImage = function AnimateImage(props) {
  var src = props.src,
      aniProps = props.aniProps,
      _props$animate = props.animate,
      animate = _props$animate === void 0 ? {
    enter: "fadeIn-300",
    leave: "fadeOut-300"
  } : _props$animate,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);

  var _useState = (0, _react.useState)(function () {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactAtomicMolecule.Image, (0, _objectSpread2["default"])({
      src: src
    }, restProps));
  }),
      image = _useState[0],
      setImage = _useState[1];

  var _mount = (0, _reshowHooks.useMounted)();

  (0, _react.useEffect)(function () {
    var oImg = new ((0, _winDoc.win)().Image)();

    oImg.onload = function () {
      if (false !== _mount()) {
        setImage( /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactAtomicMolecule.Image, (0, _objectSpread2["default"])({
          src: src
        }, restProps)));
      }
    };

    oImg.onerror = function () {
      console.warn("Get image failed. [" + src + "]");

      if (false !== _mount()) {
        setImage(null);
      }
    };

    oImg.src = src;
  }, [src]);
  return (0, _react.useMemo)(function () {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Change["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])((0, _objectSpread2["default"])({
      className: "animate-image"
    }, animate), aniProps), {}, {
      children: image
    }));
  }, [image, animate]);
};

var _default = AnimateImage;
exports["default"] = _default;
module.exports = exports.default;
