// organism-react-animate/build/cjs/src/ui/organisms/Animate.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _react = require("react");

var _reactAtomicMolecule = require("react-atomic-molecule");

var _reshowHooks = require("reshow-hooks");

var _AnimateGroup = _interopRequireDefault(require("../organisms/AnimateGroup.js"));

var _aniUtil = require("../../aniUtil.js");

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["appear", "enter", "leave"];

var Animate = function Animate(props) {
  var appear = props.appear,
      enter = props.enter,
      leave = props.leave,
      restProps = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);

  var _useState = (0, _react.useState)(false),
      isLoad = _useState[0],
      setIsLoad = _useState[1];

  var _useState2 = (0, _react.useState)({}),
      aniConf = _useState2[0],
      setAniConf = _useState2[1];

  var lastRun = (0, _react.useRef)([]);

  var _mount = (0, _reshowHooks.useMounted)();

  (0, _react.useEffect)(function () {
    var that = {};
    var data;

    var isDone = function isDone(key) {
      return function () {
        lastRun.current = lastRun.current.filter(function (val) {
          return val !== key;
        });

        if (lastRun.current.length) {
          return;
        } else {
          setTimeout(function () {
            return _mount() && setIsLoad(true);
          });
        }
      };
    };

    if (appear) {
      data = (0, _aniUtil.parseAniValue)(appear);
      that.appear = data.name;
      that.appearKey = data.key;
      that.appearTimeout = data.timeout;
      that.appearDelay = data.delay;
      that.appearClass = data.className;
      lastRun.current.push(appear);
      (0, _aniUtil.initAni)(that.appearKey, that.appear, that.appearTimeout, isDone(appear));
    }

    if (enter) {
      data = (0, _aniUtil.parseAniValue)(enter);
      that.enter = data.name;
      that.enterKey = data.key;
      that.enterTimeout = data.timeout;
      that.enterDelay = data.delay;
      that.enterClass = data.className;
      lastRun.current.push(enter);
      (0, _aniUtil.initAni)(that.enterKey, that.enter, that.enterTimeout, isDone(enter));
    }

    if (leave) {
      data = (0, _aniUtil.parseAniValue)(leave);
      that.leave = data.name;
      that.leaveKey = data.key;
      that.leaveTimeout = data.timeout;
      that.leaveDelay = data.delay;
      that.leaveClass = data.className;
      lastRun.current.push(leave);
      (0, _aniUtil.initAni)(that.leaveKey, that.leave, that.leaveTimeout, isDone(leave));
    }

    if (!appear && !enter && !leave) {
      setIsLoad(true);
    }

    setAniConf(that);
  }, [appear, enter, leave]);
  restProps.isLoad = isLoad;
  return isLoad ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_AnimateGroup["default"], (0, _objectSpread2["default"])({
    timeout: {
      appear: aniConf.appearTimeout,
      enter: aniConf.enterTimeout,
      exit: aniConf.leaveTimeout
    },
    delay: {
      appear: aniConf.appearDelay,
      enter: aniConf.enterDelay,
      exit: aniConf.leaveDelay
    },
    classNames: {
      appear: aniConf.appearClass,
      enter: aniConf.enterClass,
      exit: aniConf.leaveClass
    },
    appear: !!appear,
    enter: !!enter,
    exit: !!leave
  }, restProps)) : (0, _reactAtomicMolecule.build)(_AnimateGroup["default"])(restProps);
};

var _default = Animate;
exports["default"] = _default;
module.exports = exports.default;
