// organism-react-animate/build/cjs/src/ui/organisms/Change.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _react = require("react");

var _callFunc = _interopRequireDefault(require("call-func"));

var _reshowHooks = require("reshow-hooks");

var _Animate = _interopRequireDefault(require("../organisms/Animate.js"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["keyEqualer", "children", "onExited", "onEntered"];

var Change = function Change(props) {
  var _props$keyEqualer = props.keyEqualer,
      keyEqualer = _props$keyEqualer === void 0 ? function (item1, item2) {
    return (item1 === null || item1 === void 0 ? void 0 : item1.key) === (item2 === null || item2 === void 0 ? void 0 : item2.key);
  } : _props$keyEqualer,
      propsChildren = props.children,
      onExited = props.onExited,
      onEntered = props.onEntered,
      otherProps = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);

  var _useState = (0, _react.useState)(propsChildren),
      children = _useState[0],
      setChildren = _useState[1];

  var _mount = (0, _reshowHooks.useMounted)();

  var scheduleChildren = (0, _react.useRef)();
  var isRunning = (0, _react.useRef)(false);
  var nextCall = (0, _react.useRef)(false);
  var prevPropsChildren = (0, _reshowHooks.usePrevious)(propsChildren);

  var handleExited = function handleExited(node, isAppear) {
    if (_mount() && scheduleChildren.current) {
      setChildren(scheduleChildren.current);
      (0, _callFunc["default"])(onExited, [node, isAppear]);
    }
  };

  var handleEntered = function handleEntered(node, isAppear) {
    isRunning.current = false;

    if (_mount()) {
      (0, _callFunc["default"])(nextCall.current);
      (0, _callFunc["default"])(onEntered, [node, isAppear]);
    }
  };

  var setNext = function setNext(nextChild) {
    var reset = function reset() {
      nextCall.current = function () {
        return setNext(nextChild);
      };
    };

    if (scheduleChildren.current !== nextChild && !keyEqualer(children, nextChild)) {
      if (nextChild && !isRunning.current) {
        scheduleChildren.current = nextChild;
        nextCall.current = false;
        children ? setChildren(null) : setChildren(nextChild);
        isRunning.current = true;
      } else {
        reset();
      }
    } else {
      reset();
    }
  };

  if (prevPropsChildren !== propsChildren) {
    setNext(propsChildren);
  }

  return (0, _react.useMemo)(function () {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Animate["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, otherProps), {}, {
      onExited: handleExited,
      onEntered: handleEntered,
      children: children
    }));
  }, [children]);
};

var _default = Change;
exports["default"] = _default;
module.exports = exports.default;
