// organism-react-animate/build/cjs/src/ui/organisms/Replace.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));

var _classCallCheck2 = _interopRequireDefault(require("reshow-runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("reshow-runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("reshow-runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("reshow-runtime/helpers/inherits"));

var _createSuper2 = _interopRequireDefault(require("reshow-runtime/helpers/createSuper"));

var _defineProperty2 = _interopRequireDefault(require("reshow-runtime/helpers/defineProperty"));

var _react = require("react");

var _Change = _interopRequireDefault(require("../organisms/Change.js"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["interval"];

var Replace = /*#__PURE__*/function (_PureComponent) {
  (0, _inherits2["default"])(Replace, _PureComponent);

  var _super = (0, _createSuper2["default"])(Replace);

  function Replace() {
    var _this;

    (0, _classCallCheck2["default"])(this, Replace);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      no: 0,
      childs: {}
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleNext", function () {
      if (_this._time) {
        clearTimeout(_this._time);
      }

      var interval = _this.props.interval;

      _this.setState(function (_ref) {
        var no = _ref.no,
            childs = _ref.childs;
        no++;

        if (no >= childs.length) {
          no = 0;
        }

        return {
          no: no
        };
      });

      _this._time = setTimeout(_this.handleNext, interval);
    });
    return _this;
  }

  (0, _createClass2["default"])(Replace, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this._time) {
        clearTimeout(this._time);
      }

      var interval = this.props.interval;
      this._time = setTimeout(this.handleNext, interval);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      clearTimeout(this._time);
      this._time = null;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          interval = _this$props.interval,
          props = (0, _objectWithoutPropertiesLoose2["default"])(_this$props, _excluded);
      var _this$state = this.state,
          no = _this$state.no,
          childs = _this$state.childs;
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Change["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, props), {}, {
        children: childs[no]
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, prevState) {
      var children = nextProps.children;

      if (children === prevState.prevChildren) {
        return null;
      }

      var childs = [];

      _react.Children.map(children, function (c) {
        return c;
      }).forEach(function (child, key) {
        return childs[key] = child;
      });

      return {
        childs: childs,
        prevChildren: children
      };
    }
  }]);
  return Replace;
}(_react.PureComponent);

(0, _defineProperty2["default"])(Replace, "defaultProps", {
  interval: 5000
});
var _default = Replace;
exports["default"] = _default;
module.exports = exports.default;
