// get-window-offset/build/es/src/nearWhere.mjs
import getDomCenter from "./getDomCenter.mjs";
import getDomPositionInfo from "./getDomPositionInfo.mjs";
var getNearLocation = function getNearLocation(center, floatInfo) {
  var loc = {
    center: false,
    centerCenter: false,
    top: false,
    bottom: false,
    left: false,
    right: false
  };
  if (floatInfo.x > center.x) {
    loc.right = true;
  } else if (floatInfo.x < center.x) {
    loc.left = true;
  } else {
    loc.center = true;
  }
  if (floatInfo.y > center.y) {
    loc.bottom = true;
  } else if (floatInfo.y < center.y) {
    loc.top = true;
  } else {
    loc.center = true;
  }
  if (loc.center) {
    if (!loc.top && !loc.bottom && !loc.left && !loc.right) {
      loc.centerCenter = true;
    }
  }
  return loc;
};
var nearWhere = function nearWhere(targetEl, floatElOrFloatXY) {
  var tarCenter = getDomCenter(targetEl);
  var floatXY;
  if (floatElOrFloatXY.nodeName) {
    var _getDomPositionInfo;
    var floatElInfo = ((_getDomPositionInfo = getDomPositionInfo(floatElOrFloatXY)) === null || _getDomPositionInfo === void 0 ? void 0 : _getDomPositionInfo.domInfo) || {
      top: 0,
      left: 0
    };
    floatXY = {
      x: floatElInfo.left,
      y: floatElInfo.top
    };
  }
  if (null == floatXY) {
    floatXY = floatElOrFloatXY;
  }
  return getNearLocation({
    x: tarCenter[0],
    y: tarCenter[1]
  }, floatXY);
};
export default nearWhere;
export { getNearLocation };
