// get-window-offset/build/cjs/src/nearWhere.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.getNearLocation = exports["default"] = void 0;
var _getDomCenter = _interopRequireDefault(require("./getDomCenter.js"));
var _getDomPositionInfo2 = _interopRequireDefault(require("./getDomPositionInfo.js"));
var getNearLocation = function getNearLocation(center, floatInfo) {
  var loc = {
    center: false,
    centerCenter: false,
    top: false,
    bottom: false,
    left: false,
    right: false
  };
  if (floatInfo.x > center.x) {
    loc.right = true;
  } else if (floatInfo.x < center.x) {
    loc.left = true;
  } else {
    loc.center = true;
  }
  if (floatInfo.y > center.y) {
    loc.bottom = true;
  } else if (floatInfo.y < center.y) {
    loc.top = true;
  } else {
    loc.center = true;
  }
  if (loc.center) {
    if (!loc.top && !loc.bottom && !loc.left && !loc.right) {
      loc.centerCenter = true;
    }
  }
  return loc;
};
exports.getNearLocation = getNearLocation;
var nearWhere = function nearWhere(targetEl, floatElOrFloatXY) {
  var tarCenter = (0, _getDomCenter["default"])(targetEl);
  var floatXY;
  if (floatElOrFloatXY.nodeName) {
    var _getDomPositionInfo;
    var floatElInfo = ((_getDomPositionInfo = (0, _getDomPositionInfo2["default"])(floatElOrFloatXY)) === null || _getDomPositionInfo === void 0 ? void 0 : _getDomPositionInfo.domInfo) || {
      top: 0,
      left: 0
    };
    floatXY = {
      x: floatElInfo.left,
      y: floatElInfo.top
    };
  }
  if (null == floatXY) {
    floatXY = floatElOrFloatXY;
  }
  return getNearLocation({
    x: tarCenter[0],
    y: tarCenter[1]
  }, floatXY);
};
var _default = nearWhere;
exports["default"] = _default;
