// organism-react-ajax/build/cjs/src/ui/organisms/AjaxForm.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");
var _interopRequireWildcard = require("reshow-runtime/helpers/interopRequireWildcard");
exports.__esModule = true;
exports["default"] = void 0;
var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutPropertiesLoose"));
var _react = _interopRequireWildcard(require("react"));
var _formSerializeJs = _interopRequireDefault(require("form-serialize-js"));
var _reshowBuild = _interopRequireDefault(require("reshow-build"));
var _callFunc = _interopRequireDefault(require("call-func"));
var _ajaxStore = _interopRequireWildcard(require("../../stores/ajaxStore.js"));
var _isRunAjax = _interopRequireDefault(require("../../isRunAjax.js"));
var _excluded = ["stop", "updateUrl", "component", "action", "afterSubmit", "beforeSubmit", "callback", "errorCallback", "path"];
var getFormAction = function getFormAction(_ref) {
  var action = _ref.action,
    path = _ref.path;
  if (action) {
    return action;
  }
  if (path) {
    var baseUrl = _ajaxStore["default"].getState().get("baseUrl");
    if (baseUrl) {
      return (0, _ajaxStore.getRawUrl)({
        path: path,
        baseUrl: baseUrl
      });
    }
  }
  return null;
};
var AjaxForm = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _props$stop = props.stop,
    stop = _props$stop === void 0 ? false : _props$stop,
    _props$updateUrl = props.updateUrl,
    updateUrl = _props$updateUrl === void 0 ? false : _props$updateUrl,
    _props$component = props.component,
    component = _props$component === void 0 ? "form" : _props$component,
    propsAction = props.action,
    afterSubmit = props.afterSubmit,
    beforeSubmit = props.beforeSubmit,
    callback = props.callback,
    errorCallback = props.errorCallback,
    path = props.path,
    rest = (0, _objectWithoutPropertiesLoose2["default"])(props, _excluded);
  var _useState = (0, _react.useState)(propsAction),
    action = _useState[0],
    setAction = _useState[1];
  (0, _react.useEffect)(function () {
    setAction(getFormAction({
      action: propsAction,
      path: path
    }));
  }, [propsAction, path]);
  var handleSubmit = (0, _react.useCallback)(function (e) {
    var _callfunc;
    if (stop) {
      return;
    }
    e.preventDefault();
    var otherParams = (_callfunc = (0, _callFunc["default"])(beforeSubmit, [e])) !== null && _callfunc !== void 0 ? _callfunc : {};
    if (false === otherParams) {
      // pause by beforeSubmit
      return false;
    }
    var formDom = e.target;
    var formAction = formDom.action || getFormAction({
      action: action,
      path: path
    });
    var formParams = (0, _formSerializeJs["default"])(formDom);
    if (formAction) {
      formDom.action = formAction;
    }
    var type;
    switch (formDom.method.toLowerCase()) {
      case "post":
        type = "ajaxPost";
        break;
      /**
       * Default method
       * https://www.w3schools.com/tags/att_form_method.asp
       */
      default:
      case "get":
        type = "ajaxGet";
        otherParams.disableAjax = !(0, _isRunAjax["default"])(props);
        otherParams.updateUrl = updateUrl;
        break;
    }
    (0, _ajaxStore.ajaxDispatch)(type, (0, _objectSpread2["default"])({
      url: formAction,
      query: formParams,
      callback: callback,
      errorCallback: errorCallback
    }, otherParams));
    (0, _callFunc["default"])(afterSubmit, [e]);
  }, [path, action, stop, callback, errorCallback, updateUrl, beforeSubmit, afterSubmit]);
  return (0, _reshowBuild["default"])(component)((0, _objectSpread2["default"])({
    ref: ref,
    action: action,
    "data-path": path,
    onSubmit: handleSubmit
  }, rest));
});
AjaxForm.displayName = "AjaxForm";
var _default = AjaxForm;
exports["default"] = _default;
module.exports = exports.default;
