// reshow-url/build/es/src/handleAnchor.mjs
import getOffset from "getoffset";
import smoothScrollTo from "smooth-scroll-to";
import { doc } from "win-doc";
var goAnchorTimer;

var urlDecode = function urlDecode(s) {
  return decodeURIComponent(s);
};

var goToAnchor = function goToAnchor(anchor) {
  return function (goAnchorDelay) {
    if (!goAnchorDelay) {
      goAnchorDelay = 0;
    }

    clearTimeout(goAnchorTimer);
    goAnchorTimer = setTimeout(function () {
      try {
        var dom = document.body.querySelector(anchor);

        if (dom) {
          var pos = getOffset(dom);
          smoothScrollTo(pos.top);
        }
      } catch (e) {}
    }, goAnchorDelay);
  };
};

var getAnchorPath = function getAnchorPath(path) {
  if (!path) {
    path = doc().URL;
  } else {
    if (0 === path.indexOf("#")) {
      path = "/" + path;
    }
  }

  var pathArr = path.split("/#/");

  if (null != pathArr[1]) {
    path = "/" + pathArr[1];
  }

  var anchor = "";
  var hashStart = path.indexOf("#");
  var anchorStart = -1 !== hashStart ? hashStart : path.indexOf("%23");

  if (-1 !== anchorStart) {
    anchor = urlDecode(path.substring(anchorStart));
    path = path.substring(0, anchorStart);
  }

  var anchorArr = anchor.split("#");
  return {
    anchor,
    path,
    anchorArr,
    lastAnchor: "#" + anchorArr[anchorArr.length - 1]
  };
};

var handleAnchor = function handleAnchor(rawPath) {
  return function (goAnchorDelay) {
    var {
      anchor,
      path
    } = getAnchorPath(rawPath);

    if (anchor) {
      goToAnchor(anchor)(goAnchorDelay);
    }

    if (-1 !== path.indexOf("?")) {
      return path.split("?")[0];
    } else {
      return path;
    }
  };
};

var disableHandleAnchor = function disableHandleAnchor(path) {
  return function () {
    return path;
  };
};

export { goToAnchor, disableHandleAnchor, getAnchorPath };
export default handleAnchor;
